File: 7/shortest_completing_word.py
from collections import Counter


class Solution:
    def shortestCompletingWord(self, licensePlate, words):
        license_lower = licensePlate.lower()
        license_letter_only = ""
        for c in license_lower:
            if 'a' <= c <= 'z':
                license_letter_only += c
        license_counter = Counter(license_letter_only)

        res = 'x' * 15
        for w in words:
            word_counter = Counter(w)
            match = True
            for k in license_counter:
                if k not in word_counter or word_counter[k] < license_counter[k]:
                    match = False
                    break
            if match and len(w) < len(res):
                res = w
        return res


s = Solution()
print(s.shortestCompletingWord("1s3 PSt", ["step", "steps", "stripe", "stepple"]))

File: 7/shortest_completing_word.go
package main

import (
	"fmt"
	"strings"
)

func counter(word string) map[rune]int {
	res := make(map[rune]int)
	for _, rune := range word {
		res[rune]++
	}
	return res
}

func shortestCompletingWord(licensePlate string, words []string) string {
	licenseLower := strings.ToLower(licensePlate)
	var licenseLetterOnly []rune
	for _, c := range licenseLower {
		if c >= 'a' && c <= 'z' {
			licenseLetterOnly = append(licenseLetterOnly, c)
		}
	}
	licenseCounter := counter(string(licenseLetterOnly))

	res := "abcdeabcdeabcde"
	for _, w := range words {
		wordCounter := counter(w)
		match := true
		for k, v := range licenseCounter {
			if wordCounter[k] < v {
				match = false
				break
			}
		}
		if match && len(w) < len(res) {
			res = w
		}
	}
	return res
}

func main() {
	fmt.Println(shortestCompletingWord("1s3 PSt", []string{"step", "steps", "stripe", "stepple"}))
}
